### PR TITLE
Enable signaling of bridge errors

### PR DESCRIPTION
**Pull request by V02460**
Adds the ability of signaling permanent errors occuring at the bridge to the SDK as well as automatic sending of these events. If the bridge rejects a room event that was handed to it, the room will automatically be notified of this error.

Mainly two things are added by this PR:
- A new method `signalBridgeError` is added to the `Intent` class.
- Rejection of requests provided to the `onEvent` callback will trigger the new intent.

Bridges should not have to do much work or work at all to profit from the new feature. Every bridge rejecting the request object delivered to it (as should be the case already) will profit. If a bridge swallows errors and simply logs them, code has to be changed to reject the request object properly.

In addition to these changes there is some refactoring and documentation going on, mainly related to the `EventQueue` class.

### CODE REVIEW

**Comment by Half-Shot** [comment]
Thanks for submitting the PR. Could you give a rough changelog in the PR description for the viewers at home?

File: lib/bridge.js [hunk 1]
```diff
@@ -972,57 +963,101 @@ function queueAlgorithm(event) {
...
-        self.consumeFn(null, entry.data);
-        self._takeNext(identifier);
-    }, function(e) {
-        self.consumeFn(e, null);
-        self._takeNext(identifier);
-    });
-};
+}
+
+class EventQueuePerRoom extends EventQueue {
```
Comment (Half-Shot) [inline lib/bridge.js]: Thank you for using ES6 classes, it's quite refreshing :). Could the EventQueue stuff be taken to a seperate EventQueue.js file, as it makes for easier reviewing

File: lib/bridge.js [hunk 2]
```diff
@@ -964,29 +969,33 @@ function queueAlgorithm(event) {
 }
 
 class EventQueue {
```
Comment (Half-Shot) [inline lib/bridge.js]: For each of these classes it would be good to have a jsdoc explaining what they do.

File: lib/bridge.js [hunk 3]
```diff
@@ -964,29 +969,33 @@ function queueAlgorithm(event) {
...
+            //  events: [ {dataReady: } ],
             //  consuming: true|false
             // }
         };
         this.consumeFn = consumeFn;
     }
 
-    push(event, data, promise) {
-        var identifier = this.type === "per_room" ? event.room_id : "none";
+    push(event, dataReady) {
```
Comment (Half-Shot) [inline lib/bridge.js]: Same for each of these functions, a [JSDoc](https://en.wikipedia.org/wiki/JSDoc) line would go a long way :)

File: lib/bridge.js [hunk 4]
```diff
@@ -773,52 +783,47 @@ Bridge.prototype._onEvent = function(event) {
...
-    }
+        context: c
+    }));
 
     if (this.opts.queue.perRequest) {
-        promise = Promise.settle([
-            promise,
-            this._prevRequestPromise
-        ]);
+        dataReady = prolong(dataReady, this._prevRequestPromise)
```
Comment (Half-Shot) [inline lib/bridge.js]: Out of interest, is there anytthing wrong with `dataReady = this._prevRequestPromise.finally(() => dataReady)`?

`finally` will fire no matter the outcome, and if the promise has already resolved/rejected, it will still fire.

File: lib/bridge.js [hunk 5]
```diff
@@ -773,52 +783,47 @@ Bridge.prototype._onEvent = function(event) {
...
-    this._queue.push(event, data, promise);
+    this._queue.push(event, dataReady);
     this._queue.consume();
     return request.getPromise();
 };
 
 Bridge.prototype._onConsume = function(err, data) {
-    if (!err) {
-        this.opts.controller.onEvent(data.request, data.context);
+    if (err) {
```
Comment (Half-Shot) [inline lib/bridge.js]: Do we not need to check for `this.opts.controller.onLog` ?

File: lib/bridge.js [hunk 6]
```diff
@@ -1000,20 +1009,15 @@ class EventQueue {
     };
 
     _takeNext(identifier) {
-        var self = this;
         var events = this._queues[identifier].events;
```
Comment (Half-Shot) [inline lib/bridge.js]: Cheeky request: Make these `const`/`let`?

File: lib/bridge.js [hunk 7]
```diff
@@ -29,6 +29,16 @@ const INTENT_CULL_CHECK_PERIOD_MS = 1000 * 60; // once per minute
...
 const INTENT_CULL_EVICT_AFTER_MS = 1000 * 60 * 15; // 15 minutes
 
+/* Returns a new promise which behaves like `main`, but holds settling back
+ * until `prolonger` was settled.
+ * @param {Promise} main The promise to mimick.
+ * @param {Promise} prolonger The promise to wait for.
+ * @return {Promise} A new promise behaving like main, but waiting for prolonger.
+ */
+function prolong(main, prolonger) {
+    return Promise.all([main, prolonger.reflect()]).then(([m, _]) => m);
```
Comment (Half-Shot) [inline lib/bridge.js]: Could this not be achieved by `prolonger.finally(() => main)`?

File: lib/errors.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
+"use strict";
```
Comment (Half-Shot) [inline lib/errors.js]: Use strict is old hat at this point, you are probably safe to remove it.

File: lib/bridge.js (hunk 4 above)
Comment (V02460) [reply lib/bridge.js]: I am new to the JS Promise world. To me it looks like some flattening magic is going on in your example code, but I think it does the right thing :)

File: lib/bridge.js (hunk 5 above)
Comment (V02460) [reply lib/bridge.js]: Afaics it is accounted for in the constructor:

https://github.com/matrix-org/matrix-appservice-bridge/blob/f5fbf9d9c3f36dbad731f6151e234f4ccd6ce705/lib/bridge.js#L156-L163

**Comment by V02460** [comment]
Changed PR from “draft” to “ready for review” as [Travis CI does currently not work with draft PRs](https://travis-ci.community/t/draft-pull-requests-not-being-built/2434).

File: lib/bridge.js (hunk 5 above)
Comment (Half-Shot) [reply lib/bridge.js]: Oh aren't I clever. Yep, that's cool.

File: lib/components/intent.js [hunk 9]
```diff
@@ -535,6 +535,49 @@ Intent.prototype.setPresence = function(presence, status_msg=undefined) {
...
+ * @return {Promise}
+ */
+Intent.prototype.signalBridgeError = function(
+    roomId,
+    eventId,
+    network_name,
+    reason,
+    affected_users
+) {
+    if (!affected_users) {
```
Comment (Half-Shot) [inline lib/components/intent.js]: Interesting. Could we save some bytes here and have affected users be `["*"]`. It feels a bit nicer than trying to fetch all the current users.

File: lib/components/intent.js [hunk 10]
```diff
@@ -535,6 +535,49 @@ Intent.prototype.setPresence = function(presence, status_msg=undefined) {
...
+    }
+
+    return this.sendEvent(
+        roomId,
+        "m.room.bridge_error",
+        {
+            "network_name": network_name,
+            "reason": reason,
+            "affected_users": affected_users,
+            "m.relates_to": {
```
Comment (Half-Shot) [inline lib/components/intent.js]: +1 for relates to! :)

File: lib/components/intent.js [hunk 11]
```diff
@@ -535,6 +535,49 @@ Intent.prototype.setPresence = function(presence, status_msg=undefined) {
...
+    affected_users
+) {
+    if (!affected_users) {
+        const room = this.client.getRoom(roomId)
+        affected_users = room.getJoinedMembers().map(m => m.userId)
+    }
+
+    return this.sendEvent(
+        roomId,
+        "m.room.bridge_error",
```
Comment (Half-Shot) [inline lib/components/intent.js]: When a event type is unspecced and in a testing state, we prefer to use a custom namespace like `com.example.room.bridge_error` so it doesn't clash with any future event types. You should do this here.

Comment (V02460) [reply lib/components/intent.js]: Do you mean `com.example` literally or should I use my own domain name?

Comment (Half-Shot) [reply lib/components/intent.js]: Use your own.

File: spec/unit/intent.spec.js [hunk 12]
```diff
@@ -19,7 +19,7 @@ describe("Intent", function() {
         var clientFields = [
             "credentials", "joinRoom", "invite", "leave", "ban", "unban",
             "kick", "getStateEvent", "setPowerLevel", "sendTyping", "sendEvent",
-            "sendStateEvent", "setDisplayName", "setAvatarUrl"
+            "sendStateEvent", "setDisplayName", "setAvatarUrl", "getRoom"
```
Comment (Half-Shot) [inline spec/unit/intent.spec.js]: where is getRoom used?

File: lib/errors.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    constructor(...args) {
+        default_message(args, "The event is not known to the bridge");
+        super(...args);
+        this.name = "EventUnknownError";
+        this.reason = "m.event_unknown";
+    }
+}
+
+module.exports = {
+    wrap,
```
Comment (Half-Shot) [inline lib/errors.js]: `wrap` might be too generic a name :/

File: lib/errors.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        args[0] = msg;
+   }
+}
+
+/**
+ * Base Error for when the bride can not handle the event.
+ */
+class EventNotHandledError extends Error {
+    constructor(...args) {
+        default_message(args, "The event could not be handled by the bridge");
```
Comment (Half-Shot) [inline lib/errors.js]: You can also do `args[0] = args[0] || "This is my default message"`

The `||` operator is magical :)

File: lib/errors.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        append_msg = old_error.toString();
+    }
+    new_error.message += ":\n" + append_msg;
+    return new_error;
+}
+
+/**
+ * Sets the default message for the given error arguments.
+ */
+function default_message(args, msg) {
```
Comment (Half-Shot) [inline lib/errors.js]: As per the other comment, we can drop this.

File: lib/errors.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
+/**
+ * Append the old error message to the new one and keep its stack trace.
+ * Example:
+ *     throw wrap(e, HighLevelError, "This error is more specific");
+ */
+function wrap(
+    old_error,
```
Comment (Half-Shot) [inline lib/errors.js]: We should lint this, but var names should be camelCase

File: lib/components/intent.js [hunk 17]
```diff
@@ -535,6 +535,47 @@ Intent.prototype.setPresence = function(presence, status_msg=undefined) {
...
+    eventId,
+    network_name,
+    reason,
+    affected_users
+) {
+    return this.sendEvent(
+        roomId,
+        "de.nasnotfound.bridge_error",
+        {
+            "network_name": network_name,
```
Comment (Half-Shot) [inline lib/components/intent.js]: You don't need to quote JSON in Javascript, it's better without. Also you can use shorthand here for nearly all of these names like:

```
{
   network_name,
   reason,
   affected_users, //NOTE: You may want to validate this one
   "m.relates_to": { stuff in here ...}
}
```

File: lib/bridge.js [hunk 18]
```diff
@@ -1074,7 +1056,7 @@ BridgeContext.prototype.get = function(roomStore, userStore) {
         if (mxSender) {
             self.senders.matrix = mxSender;
         }
-    });
+    }).catch(e => {throw wrap(e, Error, "Could not retrieve bridge context");});
```
Comment (Half-Shot) [inline lib/bridge.js]: My advice for this feature is to disable it by default and stick a config flag in the constructor at the top like we do for things like room upgrades. It means bridges that rely on previous behaviour will not explode, but it will still allow them to opt in with a one liner :)

File: lib/bridge.js [hunk 19]
```diff
@@ -773,61 +776,92 @@ Bridge.prototype._onEvent = function(event) {
...
+        this._getUserRegex()
     );
 };
 
+/**
+ * Returns a regex matching all users of the bridge.
+ *
+ * @return {string} Super regex composed of all user regexes.
+ */
+Bridge.prototype._getUserRegex = function() {
```
Comment (Half-Shot) [inline lib/bridge.js]: I'd expect a RegExp object here :)

File: lib/bridge.js [hunk 20]
```diff
@@ -773,61 +776,92 @@ Bridge.prototype._onEvent = function(event) {
...
+    const contextReady = context.get(this._roomStore, this._userStore)
+
+    return contextReady.return(context);
+}
+
+Bridge.prototype._handleEventError = function(event, error) {
+    if (!error instanceof EventNotHandledError) {
+        error = wrap(error, InternalError);
+    }
+    this._botIntent.signalBridgeError(
```
Comment (Half-Shot) [inline lib/bridge.js]: Do we need to handle cases where the bot may not be in the room?

File: lib/bridge.js [hunk 21]
```diff
@@ -773,61 +776,92 @@ Bridge.prototype._onEvent = function(event) {
...
+        this.opts.controller.onLog("onEvent failure: " + err);
         return;
     }
-    if (!this.opts.controller.onLog) {
-        return;
+
+    this.opts.controller.onEvent(data.request, data.context);
+};
+
+Bridge.prototype._getBridgeContext = function(event) {
```
Comment (Half-Shot) [inline lib/bridge.js]: async function?

File: lib/bridge.js [hunk 22]
```diff
@@ -773,61 +776,92 @@ Bridge.prototype._onEvent = function(event) {
...
-        room: event.room_id
-    });
-    var data = {
+    const request = this._requestFactory.newRequest({ data: event });
+    const contextReady = this._getBridgeContext(event);
+    const dataReady = contextReady.then(c => ({
         request: request,
-        context: context
-    };
+        context: c
```
Comment (Half-Shot) [inline lib/bridge.js]: Eek, I'd rather you used shorthand here so ```{request, context}```

File: spec/unit/intent.spec.js (hunk 12 above)
Comment (V02460) [reply spec/unit/intent.spec.js]: Nowhere – it's a leftover from the old implementation.

File: lib/errors.js (hunk 14 above)
Comment (V02460) [reply lib/errors.js]: Oh…we're not doing Python here :smile: Python would make you suffer for that^^°

I'd argue to keep the method, as it makes the intent clearer. (I would still clean up `default_message` then.)

File: lib/bridge.js (hunk 22 above)
Comment (V02460) [reply lib/bridge.js]: The more you know :rainbow:

File: spec/unit/intent.spec.js (hunk 12 above)
Comment (Half-Shot) [reply spec/unit/intent.spec.js]: Can probably drop this then :)

File: lib/bridge.js (hunk 18 above)
Comment (V02460) [reply lib/bridge.js]: I am not sure which feature you are talking about here. With this line an Error is thrown, without this line an Error is thrown :thinking: Could you please elaborate a bit more?

File: lib/bridge.js (hunk 21 above)
Comment (V02460) [reply lib/bridge.js]: Well, it does return a Promise… If I added the `asyc` keyword, wouldn't that make it return `Promise<Promise<BridgeContext|null>>`?

File: lib/bridge.js (hunk 20 above)
Comment (V02460) [reply lib/bridge.js]: Yes. As this is more a design question and currently discussed in context of the MSC I might want to delay it here.

File: lib/bridge.js (hunk 19 above)
Comment (V02460) [reply lib/bridge.js]: You suggest changing the name to something like `_getUserRegexString`?

File: lib/errors.js (hunk 16 above)
Comment (V02460) [reply lib/errors.js]: https://github.com/matrix-org/matrix-appservice-bridge/pull/115

File: lib/bridge.js (hunk 18 above)
Comment (Half-Shot) [reply lib/bridge.js]: Yep, sorry. It was tricky to find the best line to use for context.

Automatically sending signal events into the room upon failure is probably not something we want on by default (which is what I assume this line is doing?). The reason for this is that it might do unexpected and unwanted things for users who are just upgrading their bridge version. Could we put automatic signaling behind a constructor flag?

File: lib/bridge.js (hunk 19 above)
Comment (Half-Shot) [reply lib/bridge.js]: I mean that this function should return a RegExp object, not a string.

File: lib/bridge.js (hunk 20 above)
Comment (Half-Shot) [reply lib/bridge.js]: You can put in a TODO line here just to explain that it's in progress.

File: lib/bridge.js (hunk 21 above)
Comment (Half-Shot) [reply lib/bridge.js]: JS is smart, and it doesn't wrap promises like that but chains them (so you will never have to handle a promise inside a promise). Making the func `async` means you can `return null;` rather than `return Promise.resolve(null);`

File: lib/bridge.js (hunk 19 above)
Comment (V02460) [reply lib/bridge.js]: Just to turn it back into a string? O_o The method is only there to get the string representation of the regex for the event.

File: lib/bridge.js (hunk 18 above)
Comment (V02460) [reply lib/bridge.js]: Oh, I get it, you are talking about this whole feature :D

https://github.com/matrix-org/matrix-appservice-bridge/blob/dcd323e17a6d0314f09756c21f0a90232b538601/lib/bridge.js#L794

This is your feature switch, essentially. Only if the bridge rejects with an `EventNotHandledError` or one of its subtypes, the behavior is triggered. By that, bridges which do not change anything, don't get anything new.

The specific line referenced by you is pretty close to a no-op: In the case of an Error already thrown, it catches it and rethrows a new one, just with another error message.

File: lib/bridge.js (hunk 19 above)
Comment (Half-Shot) [reply lib/bridge.js]: My bad, you're not using it as a regex anywhere here but just to stick inside the event. Got it.

File: lib/bridge.js (hunk 18 above)
Comment (Half-Shot) [reply lib/bridge.js]: That's very clever and I totally misunderstood it. Makes more sense to me now :)

**Comment by Half-Shot** [comment]
So this all looks good after https://github.com/matrix-org/matrix-appservice-bridge/pull/111#discussion_r304776393 gets resolved. Can we get a quick how-to use section in README.md like we do for other components so other bridge developers can understand? This should supplement the jsdoc

File: README.md [hunk 23]
```diff
@@ -179,3 +179,22 @@ will not be emitted by `onEvent`. For more information, see the docs.
...
  * `RemoteUser` - A representation of a third-party user.
+
+
+## Signaling Bridge Errors
+This section applies when you are using `Bridge` and want to notify your users
+about problems while processing their events.
+
+One thing the bridge requires you to do is fulfilling or rejecting the
+`request` promise which is handed to you as argument of the
+`controller.onEvent` callback. When rejecting the promise the `Error` you
```
Comment (Half-Shot) [inline README.md]: comma

```suggestion
`controller.onEvent` callback. When rejecting the promise, the `Error` you
```

File: README.md [hunk 24]
```diff
@@ -179,3 +179,22 @@ will not be emitted by `onEvent`. For more information, see the docs.
...
+reject with does matter:
```
Comment (Half-Shot) [inline README.md]: Matter felt a little...ambiguous?

```suggestion
reject with will indicate to the bridge library how to behave:
```

File: README.md [hunk 25]
```diff
@@ -179,3 +179,22 @@ will not be emitted by `onEvent`. For more information, see the docs.
...
+
+- On most `Error`s the bridge will simply notify the homeserver that the event
```
Comment (Half-Shot) [inline README.md]: I would reorder this list, and rephrase this to say "On all other Error types, the bridge will..."

File: README.md [hunk 26]
```diff
@@ -179,3 +179,22 @@ will not be emitted by `onEvent`. For more information, see the docs.
...
+  could not be handled right now. The homeserver will then try to resend the
+  message again with exponential backoff.
```
Comment (Half-Shot) [inline README.md]: This sounds strange to me. Would an Error cause the bridge to emit a non-200 status code to the homeserver in order to make the homeserver resend it. It sounds dangerous to me to start feeding bridges the same events, as I believe there may be bridges at present which return `Error` for minor things, while the event may have been processed.

I must have missed this in my previous review, but I would urge that Errors are handled the same way as successes and just use your custom Error types to influence bridge logic.

Comment (V02460) [reply README.md]: Reread some docs and code. I got that part wrong >.< The AS (matrix-appservice-node) will always respond with a 200 as soon as the callback finishes (regardless of what its returned promise is). The event is then irrevocable the problem of the bridge.

That makes me want to reconsider some things about the MSC. Even with that misunderstanding, I think the code is still ok in this form for now.

Comment (Half-Shot) [reply README.md]: >Reread some docs and code. I got that part wrong >.< The AS (matrix-appservice-node) will always respond with a 200 as soon as the callback finishes (regardless of what its returned promise is). The event is then irrevocable the problem of the bridge.

Ah good. I was worried you may have changed this and I missed it. Anyway that's cool. I guess you can update this README statement and then the code should be good :)

File: lib/errors.js (hunk 13 above)
Comment (V02460) [reply lib/errors.js]: https://github.com/Half-Shot/matrix-appservice-discord/pull/512#discussion_r304373100

**Comment by Half-Shot** [comment]
Looks like there are a few linting warnings we can clean up, and istanbul is not pleased about our async syntax :(

**Comment by V02460** [comment]
I cleaned up linter warnings in all files I touched and did https://github.com/matrix-org/matrix-appservice-bridge/pull/116 for the issue with Istanbul.

File: lib/bridge.js [hunk 27]
```diff
@@ -788,66 +791,97 @@ Bridge.prototype._onEvent = function(event) {
...
+        request
+        .getPromise()
+        .catch(EventNotHandledError, e => this._handleEventError(event, e))
+    );
+};
+
+/**
+ * Returns a promise that is like `promise` but adheres to all necessary limits.
+ */
+Bridge.prototype._limited = function(promise, request) {
```
Comment (Half-Shot) [inline lib/bridge.js]: Is this always returning a promise?

File: lib/bridge.js [hunk 28]
```diff
@@ -788,66 +791,97 @@ Bridge.prototype._onEvent = function(event) {
...
 
+    return (
+        request
+        .getPromise()
+        .catch(EventNotHandledError, e => this._handleEventError(event, e))
+    );
+};
+
+/**
+ * Returns a promise that is like `promise` but adheres to all necessary limits.
```
Comment (Half-Shot) [inline lib/bridge.js]: This needs to be less vague. What are "all necessary limits"?

File: lib/components/event-queue.js [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+ * Dummy EventQueue for which no queue is utilized.
+ *
+ * Every event is handled as soon as its data is available.
+ */
+class EventQueueNone extends EventQueue {
+    constructor(consumeFn) {
+        super("none", consumeFn);
+    }
+
+    push(event, data, promise) {
```
Comment (Half-Shot) [inline lib/components/event-queue.js]: as far as I can tell, you never pass a third value into this function anywhere else in the PR which makes it go bang.

File: lib/bridge.js (hunk 27 above)
Comment (V02460) [reply lib/bridge.js]: Made it async :)

File: lib/bridge.js (hunk 28 above)
Comment (V02460) [reply lib/bridge.js]: Hope its clearer now.

**Comment by Half-Shot** [comment]
Okay, have tested this on matrix-appservice-irc and it does the right thing so I'll merge it in \o/
